contracts/src/lib/SupplySourceHelper.sol

```diff
@@ -101,13 +102,36 @@ library SupplySourceHelper {
             // Use the optimized path to send value along with the call.
             (success, ret) = functionCallWithValue({target: target, data: data, value: value});
         } else if (supplySource.kind == SupplyKind.ERC20) {
-            // Transfer the tokens first, _then_ perform the call.
-            (success, ret) = ierc20Transfer(supplySource, target, value);
-            if (success) {
-                // Perform the call only if the ERC20 was successful.
-                (success, ret) = functionCallWithValue(target, data, 0);
+            (success, ret) = functionCallWithERC20Value({supplySource: supplySource, target: target, data: data, value: value});
+        }
+        return (success, ret);
+    }
+
+    /// @dev Performs the function call with ERC20 value atomically
+    function functionCallWithERC20Value(
+        SupplySource memory supplySource,
+        address target,
+        bytes memory data,
+        uint256 value
+    ) internal returns (bool success, bytes memory ret) {
+        // Transfer the tokens first, _then_ perform the call.
+        (success, ret) = ierc20Transfer(supplySource, target, value);
+
+        if (success) {
+            // Perform the call only if the ERC20 was successful.
+            (success, ret) = functionCallWithValue(target, data, 0);
+        }
+
+        if (!success) {
+            // following the implementation of `openzeppelin-contracts/utils/Address.sol`
+            if (ret.length > 0) {
+                assembly {
+                    let returndata_size := mload(ret)
+                    // see https://ethereum.stackexchange.com/questions/133748/trying-to-understand-solidity-assemblys-revert-function
+                    revert(add(32, ret), returndata_size)
+                }
             } else {
-                return (success, ret);
+                revert FailedInnerCall();
             }
```

Comment:
Let's remove the `else` and simply use `revert()`. If the target didn't provide an error, we should not add one into the mix; we're simply interested in passing through here, right?